Step2: 문자열 사칙연산 계산기 구현 및 테스트 작성

Dev-RubinJo opened the pull request:
2단계 문자열 사칙 연산 계산기를 구현하고, 그에 따른 테스트코드를 작성해봤습니다.

Dev-RubinJo committed:
refactor: 피드백 적용(불필요 메서드 제거)

Dev-RubinJo committed:
step2: 문자열 사칙연산 계산기 구현 및 테스트

wonhee009 commented on RacingCar/StringCalculator/StringCalculatorError.swift:
```diff
@@ -0,0 +1,31 @@
+    
+    var localizedDescription: String {
+        switch self {
+        case .passNil:
+            return "입력값이 nil입니다."
+        case .passEmpty:
+            return "입력값이 비어있습니다."
```
`pass`는 어떤 의미로 사용하신건가요?.?

Dev-RubinJo replied:
`pass`는 각각 nil과 빈값이 넘어왔다는 의미로 사용했습니다.

wonhee009 replied:
`pass`보다는 `isNil`, `isEmpty`는 어떠신가요?.?
`StringCalculatorError.passNil`보다는 `StringCalculatorError.isNil`이 좀 더 의미 전달이 잘 되지 않나 생각합니다 😄

Dev-RubinJo replied:
is네이밍은 bool 프로퍼티 값을 정의할때 많이 사용하다보니 에러를 정의하면서 생각을 못했던 것 같습니다. 에러에서도 is 네이밍이 좀 더 적절해보이는군요. 감사합니다!

wonhee009 commented on RacingCar/StringCalculator/StringCalculator.swift:
```diff
@@ -0,0 +1,136 @@
+    
+    private let basicOperators: [String] = ["+", "-", "*", "/"]
+    
+    mutating func runWithUserInput() throws -> Int {
+        clearData()
+        try split(mathExpression: input())
+        guard let result = operate() else {
+            throw StringCalculatorError.calculateError
+        }
+        return result
+    }
+    
+    mutating func runWithTest(expression: String?) throws -> Int {
+        clearData()
+        try split(mathExpression: expression)
+        guard let result = operate() else {
+            throw StringCalculatorError.calculateError
+        }
+        return result
+    }
```
실제 사용하기 위한 인터페이스와 테스트용 인터페이스를 분리해 놓으셨군용.
expression을 받고, 안 받고의 차이가 있을 뿐 내부 로직은 동일한 것으로 보입니다.
테스트 코드에서뿐만 아니라 실제 프로젝트에서도 
```swift
StringCalculator().runWithUserInput()
StringCalculator().runWithTest(expression: "--")
```
둘 다 사용이 가능할거 같은데요.
실제 프로젝트에서 `runWithTest`를 호출한다던지, 테스트 프로젝트에서 `runWithUserInput`을 호출한다던지 하는 실수가 있을 수 있을거 같습니다.
또한, 현재는 내부 로직이 동일하다고 해도 이후에 로직이 변경되었을 경우 `runWithUserInput` 내부만 수정한다던지 하는 실수도 있으거 같고요.

Dev-RubinJo replied:
만들때는 못느꼈지만 지금 다시 보니 완전 동일한 로직을 두번 작성해뒀군요. 하나로 통일시켜 적용할 수 있는 범위니 적용시켜두도록 하겠습니다.

wonhee009 commented on RacingCar/StringCalculator/StringCalculator.swift:
```diff
@@ -0,0 +1,136 @@
+
+import Foundation
+
+struct StringCalculator {
```
`struct`로 정의한 이유가 있을까요?.?

Dev-RubinJo replied:
계산한다는 행위를 하는 객체가 힙 메모리에 상주하여 수행하는게 적절해보이지 않는다 생각을 했습니다. 스택 메모리에서 생성되고 사라지면서 계산을 수행하는게 개발 당시에 올바르다고 생각했습니다.

wonhee009 replied:
> 계산한다는 행위를 하는 객체가 힙 메모리에 상주하여 수행하는게 적절해보이지 않는다 생각

이렇게 생각한 이유가 있을까요?

Dev-RubinJo replied:
아무래도 계산을 하는 행위가 한번 하고 나면 끝나는 것이라고 생각했습니다. 그래서 힙 메모리에 상주보다는 사용할때만 스택메모리에 추가되고, 끝나면 빠지는 것으로 생각을 했습니다.

wonhee009 replied:
"힙 메모리에 상주" 이 부분이 이해가 가지 않는데요. struct나 class가 deinit되는 시점은 현재 동일하지 않나요?
이 부분은 다음 스텝에서 좀 더 같이 얘기를 나눠보면 좋을거 같습니다.

wonhee009 commented on RacingCar/StringCalculator/StringCalculator.swift:
```diff
@@ -0,0 +1,136 @@
+    
+    private func operate() -> Int? {
+        var previous: Int?
+        var numberPointer: Int = 0, operatorPointer: Int = 0
```
`Pointer`보다는 `Index`는 어떨까요?

Dev-RubinJo replied:
생각해보니 `Index`가 더 적절할 것 같군요! 변경하도록 하겠습니다~ 당시에 `Index`가 떠오르지가 않았었네요...

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
몇몇 코멘트 남겨 두었으니 확인 부탁드려요 🙏

Dev-RubinJo committed:
fix: step2 피드백 적용

wonhee009 commented on RacingCar/StringCalculator/StringCalculatorError.swift:
```diff
@@ -0,0 +1,31 @@
+    case calculateError
+    case invalidOperator
+    
+    var localizedDescription: String {
```
[localizedError](https://developer.apple.com/documentation/foundation/localizederror)의 errorDescription 사용해 볼 수 있을거 같네요 😄

Dev-RubinJo replied:
이 부분은 생각을 못했네요. 적용해보도록 하겠습니다!

wonhee009 commented on RacingCarTest/StringCalculatorTests.swift:
```diff
@@ -0,0 +1,77 @@
+
+class StringCalculatorTests: XCTestCase {
+    
+    lazy var calculator: StringCalculator = {
+        return StringCalculator()
+    }()
+
+    override func setUpWithError() throws {
+        calculator = StringCalculator()
+    }
```
`lazy`로 정의한 이유가 있을까요?
아래의 `setUpWithError`에서 `calculator`를 초기화 해주고 있어서요.

Dev-RubinJo replied:
이게 하나의 calculator를 전역으로 가지고 테스트가 실행될때마다 새로운 인스턴스로 바꿔주는걸 의도했는데 당시에 var로만 하면 되겠다 라는 생각을 못했습니다. 지금 보면 그냥 var로 바꿔서 처리해도 제가 의도했던대로 동작이 되겠군요

Dev-RubinJo replied:
아앗 제가 아예 struct/class를 따로 분리하는 것을 생각하지 못했습니다. 말씀해주신 것을 보고 생각해보니 확실히 Calculator에 split을 하는 것은 계산한다는 행위에서 벗어나는 것이군요. 지금까지 이렇게 생각을 하지 못해서 처음 말씀해주셨던게 제대로 파악하지 못했나봅니다. 감사합니다! 한층 더 시야가 넓어진 느낌입니다. 다음 스텝에서 바로 적용해보도록 하겠습니다.

wonhee009 commented on RacingCarTest/StringCalculatorTests.swift:
```diff
@@ -0,0 +1,77 @@
+        calculator = StringCalculator()
+    }
+
+    func testAdd() throws {
+        let result = try calculator.run(with: "1 + 2")
+        XCTAssertEqual(result,
+                       3,
+                       StringCalculatorError.calculateError.localizedDescription)
+    }
```
`testAdd`라고 하기에는 calculator의 역할이 add만 하진 않는거 같습니다.
현재 calculator가 하는 일을 확인해보면 
1. 입력 받은 input을 공백을 기준으로 split
2. split된 input을 숫자와 기호로 구분
3. 분류된 숫자와 기호를 기준으로 기호 type을 판단
4. 숫자와 기호로 계산

이런 프로세스를 거치고 있는데요.
split과 계산 2가지 역할을 calculator가 가지고 있는거 같습니다.
해당 역할을 나누고 split과 계산에 대한 테스트 코드를 작성하는건 어떨까요?

Dev-RubinJo replied:
아직 함수의 분리가 정확하게 와닫지 않았었군요. run(with:) 메서드가 두가지 일을 한다는 것을 제대로 인지하지 못하고 있었습니다. 위의 테스트는 testRun이 더 적절할 것 같군요. test받는 역할에 따라 보다 더 구체적으로 나눠보도록 하겠습니다.

Dev-RubinJo committed:
feat(step2): 2차 피드백 적용

1. Error naming 수정
2. StringCalculatorError에 LocalizedError extension 적용하여 errorDescription 정의
3. StringCalculatorTests에서 lazy var로 calculator를 만들었던 것 의도에 맞게 수정
4. 하나의 메서드에 하나의 테스트를 할 수 있도록 수정

wonhee009 commented on RacingCar/StringCalculator/StringCalculator.swift:
```diff
@@ -0,0 +1,123 @@
+    
+    private let basicOperators: [String] = ["+", "-", "*", "/"]
+    
+    mutating func run(with expression: String?) throws -> Int {
+        clearData()
+        try split(mathExpression: expression)
+        guard let result = operate() else {
+            throw StringCalculatorError.calculateError
+        }
+        return result
+    }
```
`calculator`내에서 split을 하고 있는건 동일한거 같은데요.
split을 해주는 struct/class를 하나 만들어 역할을 분리하는건 어떨까요?
input을 split만 해주는 역할/계산만 해주는 역할 이렇게 분리할 수 있을거 같습니다.

wonhee009 reviewed (approved):
수고하셨습니다 🎉 
코멘트 남겨 두었으니 해당 부분은 다음 스텝에서 고려해주세요 😄 
다음 스텝에서 뵙겠습니다!
